external/fv3fit/fv3fit/emulation/thermobasis/emulator.py

```diff
@@ -0,0 +1,366 @@
+
+            if validation_data:
+                loss_epoch_test = self.score(validation_data)
+                self.log_dict("test_epoch", loss_epoch_test, step=i)
+                x, y = next(iter(validation_data.batch(3).take(1)))
+                in_ = self.batch_to_specific_humidity_basis(x)
+                out = batch_to_specific_humidity_basis(y)
+                pred = self.model(in_)
+                if isinstance(self.model, UVTQSimple):
+                    self.log_profiles(
+                        "eastward_wind_truth", (out.u - in_.u).numpy().T, step=i
+                    )
+                    self.log_profiles(
+                        "eastward_wind_prediction", (pred.u - in_.u).numpy().T, step=i
+                    )
+                    self.log_profiles(
+                        "humidity_truth", (out.q - in_.q).numpy().T, step=i
+                    )
+                    self.log_profiles(
+                        "humidity_prediction", (pred.q - in_.q).numpy().T, step=i
+                    )
+
+    def log_profiles(self, key, data, step):
+        self.logger.log_profiles(key, data, step)
+
+    def log_dict(self, prefix, metrics, step):
+        self.logger.log_dict(prefix, metrics, step)
+
+    @property
+    def _checkpoint(self) -> tf.train.Checkpoint:
+        return tf.train.Checkpoint(optimizer=self.optimizer, model=self.model)
+
+    _config = "config.yaml"
+    _model = "model"
+
+    def dump(self, path: str):
+        if path:
+            os.makedirs(path, exist_ok=True)
+        with open(os.path.join(path, self._config), "w") as f:
+            json.dump(dataclasses.asdict(self.config), f)
+        self._checkpoint.write(os.path.join(path, self._model))
+
+    @classmethod
+    def load(cls, path: str):
+        with open(os.path.join(path, cls._config), "r") as f:
+            config = Config.from_dict(json.load(f))
+        model = cls(config)
+        model._checkpoint.read(os.path.join(path, cls._model))
+        return model
+
+
+def get_model(config: Config) -> tf.keras.Model:
+    if config.cloud_water:
+        logging.info("Using V1QCModel")
+        return V1QCModel(config.levels)
+    elif isinstance(config.target, MultiVariableLoss):
+        logging.info("Using ScalerMLP")
+        n = config.levels
+        if config.relative_humidity:
+            return UVTRHSimple(n, n, n, n)
+        else:
+            return UVTQSimple(n, n, n, n)
+    elif isinstance(config.target, QVLoss):
+        logging.info("Using ScalerMLP")
+        model = ScalarMLP(
+            var_level=config.target.level,
+            num_hidden=config.num_hidden,
+            num_hidden_layers=config.num_hidden_layers,
+        )
+    elif isinstance(config.target, RHLoss):
+        logging.info("Using RHScaler")
+        model = RHScalarMLP(
+            var_level=config.target.level,
+            num_hidden=config.num_hidden,
+            num_hidden_layers=config.num_hidden_layers,
```

Comment:
If this is meant to be extended and there is a one to one relationship between loss targets and a model, I'd suggest having `Target.model` or even `Config.model` which allows dependency injection. Right now adding new models requires modifying this function.

Reply:
I agree this would be nice, but I'm okay violating the open-closed rule when it comes to these kinds of simple factories functions. If we start to heavily extend this model, I agree we will need a better plugin architecture. As long as this function fits on one screen, I would advocate for leaving it unchanged.

Reply:
Can we codify in a team meeting when we should or should not enforce OCP, and put it in our guidelines? I'm fine with this, but I've had a few PRs blocked by cases where I feel it's okay to violate the open-closed rule, and would like us to be consistent.

Reply:
I agree this is an OCP violation. Do you think it is worth addressing? I was concerned adding a registration/plugin framework would be over-engineering for this simple function. Is this function hard for you to understand? Ultimately, we could probably hook this into the fv3fit registration framework.

> I've had a few PRs blocked by cases where I feel it's okay to violate the open-closed rule, and would like us to be consistent. 

It sounds like you feel there is not consistency in our review process. The OCP is design principle, not a hard rule. Some OCP violations (e.g. changing widely-used APIs) are more potent than others. I think in a review, both the reviewer and the author have some say in whether an OCP violation is worth addressing. Sometimes a minor OCP violation (e.g. a short switch-statement in factory function) may be more complex to work around.

Reply:
> I agree this is an OCP violation. Do you think it is worth addressing?

I said I am fine with this. If this is meant to be re-used by other people for other models I'd ask for it to be addressed, but my understanding is this is to be used in this one model.

I don't think there is uniform understanding across the team about what the OCP means, and it would help to have a team discussion about what it is and when we should or shouldn't follow it. It's a good principle that the team should develop a better understanding of. Each of the [six times](https://github.com/VulcanClimateModeling/fv3net/pulls?q=%22open-closed+principle%22+in%3Acomments) the open/closed principle it has been mentioned by a reviewer, you were the reviewer. So yes, it does feel like there is inconsistency in the review process.

Reply:
To be clear, I'm not saying there's anything wrong with your reviews, I just think this is a particularly relevant case where a principle is kind of hard to understand, and not everyone in the team is taking it into account in their reviews.